### PR TITLE
feat: survey scheduling

### PR DESCRIPTION
**Pull request by xernobyl**
## What does this PR do?

Adds survey publish/pause scheduling on top of the BullMQ infrastructure.

This branch introduces end-to-end support for optional survey scheduling dates and a central BullMQ reconciliation job that applies those transitions automatically.

Fixes formbricks/internal#1558

### Summary

- Adds optional `publishOn` and `pauseOn` fields to surveys.
- Exposes scheduling controls in the survey editor `Settings` tab.
- Persists scheduling metadata through the shared survey contracts/types.
- Adds a BullMQ recurring job that reconciles due survey status transitions daily at `00:00 CET`.
- Applies due transitions immediately on create/update so users do not need to wait for the cron run when the selected date is already due.

#### 1. Data model and shared contracts

- Adds nullable `publishOn` and `pauseOn` columns to the Prisma `Survey` model.
- Adds lookup indexes on `(status, publishOn)` and `(status, pauseOn)` to support scheduler queries efficiently.
- Threads the new fields through:
  - shared Prisma survey selection / transformation
  - `@formbricks/types` survey types
  - `@formbricks/database/zod/surveys`
  - management API survey types
  - editor survey state handling

#### 2. Scheduling semantics

- Treats the picker value as a date-only selection, not a browser-local timestamp.
- Stores date selections as stable noon-UTC values client-side to avoid timezone drift.
- Normalizes the selected date server-side to fixed `00:00 CET` (`Etc/GMT-1`) before persistence.
- Uses the current fixed-CET calendar day as the minimum selectable date in the UI, so “today” is evaluated consistently even for users outside CET.

#### 3. Survey editor UI

- Adds a dedicated `Survey schedule` card in the survey editor `Settings` view.
- Supports independent `Publish on date` and `Pause on date` fields.
- Adds clear actions for both scheduled dates.
- Shows the CET helper copy:
  - `Survey will be published at 00:00 CET on the selected date`
  - `Survey will be paused at 00:00 CET on the selected date`
- Renders the card as a collapsible section:
  - collapsed by default when no schedule is set
  - auto-expanded when either date is set
  - summary lines shown in the collapsed header
- Moves the scheduling section to the bottom of the settings screen.

#### 4. Save flow and immediate reconciliation

- Centralizes scheduling normalization in a server-side scheduling helper.
- Clears only obsolete schedules on manual status changes:
  - manual publish (`draft` -> `inProgress`) clears `publishOn`
  - manual pause / completion clears `pauseOn`
- Reuses the same post-persist reconciliation path in both `createSurvey()` and `updateSurveyInternal()`.
- If a persisted survey is already due at save time, the service immediately runs the reconciliation logic and returns the final survey status instead of waiting for the daily job.

This also covers the same-day edge case:

- if `publishOn` and `pauseOn` are both due on the same day, the reconciliation order is publish first, then pause, so the final state is `paused`

Note: the product task uses “published”, but the existing Formbricks survey lifecycle uses `inProgress` as the live status. This implementation follows the existing status model.

#### 5. BullMQ integration

- Adds a dedicated BullMQ job type: `survey-scheduling.reconcile`
- Adds producer helpers in `@formbricks/jobs` for:
  - enqueueing survey scheduling jobs
  - one-off scheduling
  - recurring schedule upsert
- Adds the survey scheduling processor registration and typed job payloads.
- Implements reconciliation in two phases:
  - publish due draft surveys
  - pause due in-progress surveys
- Uses guarded `updateMany()` transitions so reruns stay idempotent and duplicate updates are avoided.

#### 6. Worker bootstrap and recurring registration

- Registers one global recurring job with:
  - `scheduleId: "daily-survey-scheduling"`
  - cron pattern `0 0 * * *`
  - timezone `Etc/GMT-1`
- Keeps recurring schedule registration separate from in-process worker startup so the schedule is also registered correctly in external-worker deployments.
- Reuses the existing startup retry pattern for both worker boot and recurring schedule registration.

#### 7. Logging and auditability

- Logs reconciliation failures and transition failures with context.
- Writes system audit log entries for automatic publish/pause transitions via `queueAuditEventWithoutRequest`.
- Does not fail an already-successful transition if audit logging itself fails; instead it logs the audit failure separately.

#### 8. Tests and localization

- Adds focused unit/service coverage for:
  - CET date normalization
  - same-save catch-up behavior
  - manual status cleanup
  - same-day publish+pause behavior
- Adds processor / queue / instrumentation coverage for the new BullMQ scheduling path.
- Adds a Playwright spec for the scheduling UI flow.
- Adds translation keys for the scheduling UI across the supported web locales.

### Manual verification

1. Start from this branch on top of `epic/bullmq`.
2. Run the database migration:
   - `pnpm db:migrate:dev`
3. Start the app and worker setup as usual for local development.
4. Create a new draft survey.
5. Open the survey editor and go to `Settings`.
6. Scroll to the bottom and expand `Survey schedule`.
7. Verify:
   - both `Publish on date` and `Pause on date` are available
   - the helper text mentions `00:00 CET`
   - both fields can be used independently
8. Set a future publish date and pause date, save, then reload the survey and confirm both dates persist.
9. Clear one or both dates and confirm the survey returns to the default unscheduled behavior.
10. Test immediate reconciliation:
    - set a due `publishOn` date on a draft survey and save
    - confirm the survey moves to `inProgress`
11. Test immediate pause reconciliation:
    - set a due `pauseOn` date on an `inProgress` survey and save
    - confirm the survey moves to `paused`
12. Test the same-day edge case:
    - set both dates to the same due day on a draft survey
    - confirm the final state is `paused`
13. Optionally verify recurring job registration in the logs for the daily survey scheduling schedule.

### Automated checks run locally

- `pnpm --filter @formbricks/web exec vitest run instrumentation.test.ts instrumentation-jobs.test.ts lib/survey/service.scheduling.test.ts modules/survey/scheduling/lib/date-utils.test.ts modules/survey/scheduling/lib/process-survey-scheduling-job.test.ts`
- `pnpm --filter @formbricks/web exec eslint instrumentation.ts instrumentation.test.ts instrumentation-jobs.ts instrumentation-jobs.test.ts lib/survey/service.ts lib/survey/service.scheduling.test.ts modules/survey/editor/components/survey-scheduling-card.tsx modules/survey/scheduling/lib/date-utils.ts modules/survey/scheduling/lib/date-utils.test.ts`
- `pnpm --filter @formbricks/web exec vitest run modules/survey/scheduling/lib/date-utils.test.ts lib/survey/service.scheduling.test.ts`
- `pnpm --filter @formbricks/web exec eslint modules/survey/scheduling/lib/date-utils.ts modules/survey/scheduling/lib/survey-scheduling.ts modules/survey/editor/components/survey-scheduling-card.tsx modules/survey/scheduling/lib/date-utils.test.ts lib/survey/service.scheduling.test.ts`
- `pnpm i18n:validate`

### Notes

- A dedicated Playwright spec was added at `apps/web/playwright/survey-scheduling.spec.ts`.
- The spec was updated to match the final collapsible scheduling UI, but I would still recommend a quick manual UI pass before merging.
- I did not run a full workspace `pnpm build` from this description workflow.

### Required

- [x] Filled out the "How to test" section in this PR
- [ ] Read [How we Code at Formbricks](<[https://github.com/formbricks/formbricks/blob/main/CONTRIBUTING.md](https://formbricks.com/docs/contributing/how-we-code)>)
- [ ] Self-reviewed my own code
- [ ] Commented on my code in hard-to-understand bits
- [ ] Ran `pnpm build`
- [ ] Checked for warnings, there are none
- [ ] Removed all `console.logs`
- [ ] Merged the latest changes from main onto my branch with `git pull origin main`
- [ ] My changes don't cause any responsiveness issues
- [ ] First PR at Formbricks? [Please sign the CLA!](https://cla-assistant.io/formbricks/formbricks) Without it we wont be able to merge it 🙏

### Appreciated

- [ ] If a UI change was made: Added a screen recording or screenshots to this PR
- [ ] Updated the Formbricks Docs if changes were necessary

### CODE REVIEW

File: apps/web/modules/survey/editor/components/survey-scheduling-card.tsx [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+              <p className="text-sm text-slate-500">
+                {t("environments.surveys.edit.schedule_survey_publish_and_pause_dates")}
+              </p>
+            )}
+          </div>
+          <ChevronDownIcon
+            className={cn("mt-1 h-5 w-5 shrink-0 text-slate-500 transition-transform", open && "rotate-180")}
+          />
+        </button>
+      </Collapsible.CollapsibleTrigger>
```
[the remark this answers is not in the document]
Comment (xernobyl) [reply apps/web/modules/survey/editor/components/survey-scheduling-card.tsx]: Addressed.

File: apps/web/modules/survey/editor/components/survey-scheduling-card.tsx [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+                  pauseOn: null,
+                }));
+              }}
+              updateSurveyDate={(date) => {
+                setLocalSurvey((currentSurvey) => ({
+                  ...currentSurvey,
+                  pauseOn: toDateOnlySelection(date),
+                }));
+              }}
+            />
```
[the remark this answers is not in the document]
Comment (xernobyl) [reply apps/web/modules/survey/editor/components/survey-scheduling-card.tsx]: Addressed.

File: apps/web/modules/survey/scheduling/lib/date-utils.test.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+  toDateOnlySelection,
+} from "./date-utils";
+
+describe("survey scheduling date utils", () => {
+  test("stores selected dates as noon UTC date-only values", () => {
+    const selectedDate = new Date(2026, 3, 17, 9, 15, 0, 0);
+    const storedDate = toDateOnlySelection(selectedDate);
+
+    expect(storedDate.toISOString()).toBe("2026-04-17T12:00:00.000Z");
+  });
```
[the remark this answers is not in the document]
Comment (xernobyl) [reply apps/web/modules/survey/scheduling/lib/date-utils.test.ts:20]: Addressed.

File: apps/web/modules/survey/scheduling/lib/date-utils.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+export const normalizeDateOnlySelectionToCETMidnight = (date: Date | null): Date | null => {
+  if (!date) {
+    return null;
+  }
+
+  const { day, month, year } = getUtcDateParts(date);
+
+  return new Date(Date.UTC(year, month, day, 0, 0, 0, 0) - FIXED_CET_OFFSET_MS);
+};
```
[the remark this answers is not in the document]
Comment (xernobyl) [reply apps/web/modules/survey/scheduling/lib/date-utils.ts:157]: Addressed.

File: apps/web/modules/survey/scheduling/lib/survey-scheduling.ts [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+      "Survey scheduling transition failed"
+    );
+
+    throw error;
+  }
+
+  await queueTransitionAudit(candidate, transition, logContext);
+
+  return true;
+};
```
[the remark this answers is not in the document]
Comment (xernobyl) [reply apps/web/modules/survey/scheduling/lib/survey-scheduling.ts:229]: Addressed.

File: apps/web/modules/ui/components/date-picker/index.tsx [hunk 6]
```diff
@@ -1,53 +1,69 @@
...
+    month: "short",
+    year: "numeric",
+  });
+};
+
+const getDefaultMinDate = (): Date => {
+  const now = new Date();
+
+  return new Date(now.getFullYear(), now.getMonth(), now.getDate(), 12, 0, 0, 0);
 };
```
[the remark this answers is not in the document]
Comment (xernobyl) [reply apps/web/modules/ui/components/date-picker/index.tsx:31]: Addressed.

File: apps/web/modules/ui/components/date-picker/index.tsx [hunk 7]
```diff
@@ -1,53 +1,69 @@
...
-  );
+  const [formattedDate, setFormattedDate] = useState<string | undefined>(getDisplayDate(date, locale));
   const [isOpen, setIsOpen] = useState(false);
-
   const btnRef = useRef<HTMLButtonElement>(null);
 
+  useEffect(() => {
+    onChange(date ? new Date(date) : undefined);
+    setFormattedDate(getDisplayDate(date, locale));
+  }, [date, locale]);
```
[the remark this answers is not in the document]
Comment (xernobyl) [reply apps/web/modules/ui/components/date-picker/index.tsx]: Addressed.

File: apps/web/playwright/survey-scheduling.spec.ts [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+  await page.getByRole("button", { name: "Pick a date" }).nth(pickerIndex).click();
+  await page
+    .locator("[data-radix-popper-content-wrapper]")
+    .last()
+    .getByRole("button", { name: new RegExp(`^${targetDate.getDate().toString()}$`) })
+    .click();
+
+  return targetDate;
+};
```
[the remark this answers is not in the document]
Comment (xernobyl) [reply apps/web/playwright/survey-scheduling.spec.ts]: Addressed.

File: apps/web/playwright/survey-scheduling.spec.ts [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+    await scheduleCardTrigger.scrollIntoViewIfNeeded();
+    await scheduleCardTrigger.click();
+
+    await expect(page.getByRole("button", { name: formatSelectedDate(publishDate) })).toBeVisible();
+    await expect(page.getByRole("button", { name: formatSelectedDate(pauseDate) })).toBeVisible();
+
+    await page.getByTestId("clear-publish-on-date").click();
+    await page.getByTestId("clear-pause-on-date").click();
+
+    await expect(page.getByRole("button", { name: "Pick a date" })).toHaveCount(2);
```
[the remark this answers is not in the document]
Comment (xernobyl) [reply apps/web/playwright/survey-scheduling.spec.ts]: Addressed.

File: packages/database/migration/20260417120000_add_survey_publish_pause_scheduling/migration.sql [hunk 10]
```diff
@@ -0,0 +1,7 @@
+ALTER TABLE "Survey"
+ADD COLUMN "publishOn" TIMESTAMP(3),
+ADD COLUMN "pauseOn" TIMESTAMP(3);
+
+CREATE INDEX "Survey_status_publishOn_idx" ON "Survey"("status", "publishOn");
+
+CREATE INDEX "Survey_status_pauseOn_idx" ON "Survey"("status", "pauseOn");
```
[the remark this answers is not in the document]
Comment (xernobyl) [reply packages/database/migration/20260417120000_add_survey_publish_pause_scheduling/migration.sql]: Addressed.

File: apps/web/modules/survey/scheduling/lib/survey-scheduling.ts [hunk 11]
```diff
@@ -0,0 +1,301 @@
...
+
+  if (isManualStatusChange && (status === "paused" || status === "completed")) {
+    normalizedPauseOn = null;
+  }
+
+  return {
+    pauseOn: normalizedPauseOn,
+    publishOn: normalizedPublishOn,
+  };
+};
```
[the remark this answers is not in the document]
Comment (xernobyl) [reply apps/web/modules/survey/scheduling/lib/survey-scheduling.ts:307]: Addressed.

**Comment by jobenjada** [comment]
image


hey @xernobyl thanks for the PR!

The initial ticket from me wasn't well written: We don't want the survey to get `paused` on date, but `completed` on date.

Also, it's not good to auto-publish surveys out of state `draft` because they might not be finished. This is why I've made the following changes:

- Changed scheduler publish eligibility from `draft` to `paused` in survey-scheduling logic.
- Updated scheduling normalization so scheduling to paused keeps closeOn when publishOn is set (avoids losing close date).
- Implemented draft schedule flow in survey-menu-bar: primary CTA becomes Schedule survey, runs publish-level validation, saves as paused, redirects with ?scheduled=true.
- Added summary success handling for scheduled flow: toast Survey scheduled successfully, no confetti, and query-param cleanup.
- Added scheduled-vs-paused UX distinctions where publishOn is already available (SurveyStatusDropdown, SurveyStatusIndicator, inactive survey UI, response card header).
- Added/updated English translation keys for scheduled labels and schedule success messaging.
- Updated regression coverage (unit + E2E specs) for new semantics; Vitest suites pass

Please review and test the changes I've made to see if they still fall in line with your initial implementation (I mostly changed UX and the field name from `pauseOn` to `closeOn`

Thanks!

**Review by jobenjada — DISMISSED** [review]
see first comment

**Comment by xernobyl** [comment]
If anyone needs to ~debug~ test this, you can change the starting time by changing

```
export const SURVEY_SCHEDULING_TIME_ZONE = "Europe/Berlin";
export const SURVEY_SCHEDULING_LOCAL_HOUR = 0;
export const SURVEY_SCHEDULING_LOCAL_MINUTE = 0;
```

In `apps/web/modules/survey/scheduling/lib/constants.ts`

File: apps/web/app/(app)/workspaces/[workspaceId]/surveys/[surveyId]/(analysis)/summary/components/SuccessMessage.tsx [hunk 12]
```diff
@@ -22,8 +22,26 @@ export const SuccessMessage = ({ survey }: SummaryMetadataProps) => {
   const appSetupCompleted = workspace.appSetupCompleted;
 
   useEffect(() => {
-    const newSurveyParam = searchParams?.get("success");
-    if (newSurveyParam && survey && workspace) {
+    const publishSuccessParam = searchParams?.get("success");
+    const scheduledSuccessParam = searchParams?.get("scheduled");
+    if (!survey || !workspace) {
```
Comment (Dhruwang) [inline apps/web/app/(app)/workspaces/[workspaceId]/surveys/[surveyId]/(analysis)/summary/components/SuccessMessage.tsx]: I think we can now remove this since survey and workspace would be always defined based on their types

File: packages/database/migration/20260420144000_rename_survey_pause_on_to_close_on/migration.sql
Comment (Dhruwang) [inline packages/database/migration/20260420144000_rename_survey_pause_on_to_close_on/migration.sql:1]: I think we don't need this sql migration, can you recheck ?

File: apps/web/lib/survey/service.ts [hunk 13]
```diff
@@ -107,74 +117,57 @@ export const selectSurvey = {
...
-  triggerIds.forEach((triggerId) => {
-    if (!actionClasses.find((actionClass) => actionClass.id === triggerId)) {
-      throw new InvalidInputError("Invalid trigger id");
-    }
-  });
-
-  if (new Set(triggerIds).size !== triggerIds.length) {
-    throw new InvalidInputError("Duplicate trigger id");
-  }
+  checkSurveyTriggersValidity(triggers, actionClasses);
```
Comment (Dhruwang) [inline apps/web/lib/survey/service.ts]: now checkTriggersValidity and handleTriggerUpdates are both just wrapper functions, can we inline them ?

File: apps/web/app/api/v2/client/[environmentId]/environment/route.test.ts [hunk 14]
```diff
@@ -148,5 +148,5 @@ describe("api/v2 client environment route", () => {
         }),
       })
     );
-  });
+  }, 20000);
```
Comment (Dhruwang) [inline apps/web/app/api/v2/client/[environmentId]/environment/route.test.ts]: any reason for increasing this timeouts ?

File: apps/web/modules/survey/components/template-list/lib/survey.ts [hunk 15]
```diff
@@ -1,14 +1,16 @@
...
 import { TSurvey, TSurveyCreateInput } from "@formbricks/types/surveys/types";
 import { getOrganization, subscribeOrganizationMembersToSurveyResponses } from "@/lib/organization/service";
 import { validateMediaAndPrepareBlocks } from "@/lib/survey/utils";
 import { getOrganizationIdFromWorkspaceId } from "@/lib/utils/helper";
-import { TriggerUpdate } from "@/modules/survey/editor/types/survey-trigger";
 import { getActionClasses } from "@/modules/survey/lib/action-class";
 import { selectSurvey } from "@/modules/survey/lib/survey";
+import { handleTriggerUpdates } from "@/modules/survey/lib/trigger-updates";
+
+export { handleTriggerUpdates } from "@/modules/survey/lib/trigger-updates";
```
Comment (Dhruwang) [inline apps/web/modules/survey/components/template-list/lib/survey.ts]: I think we wont need this export here, can you check ?

File: apps/web/modules/survey/editor/components/survey-menu-bar.tsx [hunk 16]
```diff
@@ -150,6 +150,22 @@ export const SurveyMenuBar = ({
...
+    draftPrimaryLabel = t("workspace.surveys.edit.save_and_close");
+  }
+
+  const getDraftSurveyToPersist = (draftSurvey: TSurvey, segment: TSegment | null): TSurveyDraft =>
+    ({
+      ...draftSurvey,
+      closeOn: draftSurvey.publishOn ? null : draftSurvey.closeOn,
+      publishOn: null,
+      segment,
+    }) as unknown as TSurveyDraft;
```
Comment (Dhruwang) [inline apps/web/modules/survey/editor/components/survey-menu-bar.tsx]: I think as unknown is not necessary here

File: apps/web/modules/survey/scheduling/lib/constants.ts
Comment (Dhruwang) [inline apps/web/modules/survey/scheduling/lib/constants.ts:1]: SURVEY_SCHEDULING_TIME_ZONE in apps/web/modules/survey/scheduling/lib/constants.ts is hardcoded to "Europe/Berlin". This means all survey schedules execute at midnight CET/CEST regardless of where the instance is hosted — a self-hoster in the US would see their surveys publish/close at 3-6pm the previous day. Can we make it configurable using an environment variable?

**Review by Dhruwang — CHANGES_REQUESTED** [review]
Thanks for the PR, looks good. Left some comments, Can you take a look ? 😊 

We should also add a validation here to not allow publish and close on the same date. 
Screenshot 2026-04-21 at 11 11 20 AM

Do we not want to show the scheduled status here? 
Screenshot 2026-04-21 at 12 01 20 PM

For scheduled app surveys, I don't see the status here. 
Screenshot 2026-04-21 at 12 28 30 PM

File: apps/web/modules/survey/scheduling/lib/constants.ts
Comment (xernobyl) [reply apps/web/modules/survey/scheduling/lib/constants.ts:1]: That makes sense. I've changed it, also added the time, so it doesn't necessarily run at midnight... Although in the future we should be able to trigger surveys at specific times, like after an event ends or something like that.

File: apps/web/modules/survey/editor/components/survey-menu-bar.tsx (hunk 16 above)
Comment (xernobyl) [reply apps/web/modules/survey/editor/components/survey-menu-bar.tsx]: Done.

File: apps/web/modules/survey/components/template-list/lib/survey.ts (hunk 15 above)
Comment (xernobyl) [reply apps/web/modules/survey/components/template-list/lib/survey.ts]: Fixed.

File: apps/web/app/api/v2/client/[environmentId]/environment/route.test.ts (hunk 14 above)
Comment (xernobyl) [reply apps/web/app/api/v2/client/[environmentId]/environment/route.test.ts]: I had some issues while testing. Reverting.

File: apps/web/lib/survey/service.ts (hunk 13 above)
Comment (xernobyl) [reply apps/web/lib/survey/service.ts]: Done.

File: packages/database/migration/20260420144000_rename_survey_pause_on_to_close_on/migration.sql
Comment (xernobyl) [reply packages/database/migration/20260420144000_rename_survey_pause_on_to_close_on/migration.sql:1]: Yes, left over from the back and forward...

File: apps/web/app/(app)/workspaces/[workspaceId]/surveys/[surveyId]/(analysis)/summary/components/SuccessMessage.tsx (hunk 12 above)
Comment (xernobyl) [reply apps/web/app/(app)/workspaces/[workspaceId]/surveys/[surveyId]/(analysis)/summary/components/SuccessMessage.tsx]: Fixed.

**Review by Dhruwang — APPROVED** [review]
Thanks for the changes, looks good 🚀
